[release-builder] local simulation of governance proposals

vgao1996 opened the pull request:
## Description
This introduces a new release builder command that enables the simulation of governance proposals. Currently only multi-step proposals are supported.

It utilizes the the remote debugger infrastructure to fetch real chain states for local simulation, but adds another in-memory database to store the new side effects generated by the governance scripts.

Normally, governance scripts needs to be approved through on-chain governance before they could be executed. This process involves setting up various states (e.g., staking pool, delegated voter), which can be quite complex.

This simulation bypasses these challenges by patching specific Move functions with mock versions, most notably `fun resolve_multi_step_proposal`, thus allowing the governance process to be skipped altogether. In other words, this simulation is intended for checking whether a governance proposal will execute successfully, assuming it gets approved.

## How to run simulation
First generate the proposal
```
cargo run -p aptos-release-builder generate-proposals --release-config data/release.yaml --output-dir 
output
```
Then run simulation via the following command
```
cargo run -p aptos-release-builder simulate-multi-step-proposal --network mainnet --proposal-dir output/sources/v1.14/step_1_upgrade_framework/
```

Here's how the output should look like
```
Found 2 scripts
    output/sources/v1.14/step_1_upgrade_framework/0-gas-schedule.move
    output/sources/v1.14/step_1_upgrade_framework/1-features.move
Compiling scripts...
Compiling, may take a little while to download git dependencies...
INCLUDING DEPENDENCY AptosFramework
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING script
Compiling, may take a little while to download git dependencies...
INCLUDING DEPENDENCY AptosFramework
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING script
Patching framework functions to bypass governance.. done
Creating and funding sender account.. done
Executing governance scripts...
    0-gas-schedule.move
        Keep(
            Success,
        )
    1-features.move
        Keep(
            Success,
        )
All scripts succeeded!
```

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Other (release tooling)

## Key Areas to Review
`simulate.rs`

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+//! This module implements the simulation of governance proposals.
+//! Currently, it supports only multi-step proposals.
+//!
+//! It utilizes the the remote debugger infrastructure to fetch real chain states
```
typo: the the

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+ * Compiled Module Helpers
+ *
+ **************************************************************************************************/
+fn find_function_def_by_name<'a>(
```
I am wondering if these kinds of helpers can be useful even in `third_party`?

vgao1996 replied:
They could be. Although I'm thinking that maybe we should refine the APIs a little bit if we want to offer them as a general solution. Might also be a good idea to see @runtian-zhou's work on the intent engine as it ships with a script builder, and see if we can make some good unified helpers out of both use cases.

runtian-zhou replied:
Yea I actually have to implement those helpers in my intent builder as well. Let's not worry about them now and we can clean this logic up later.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+    }
+}
+
+fn get_or_add_addr(m: &mut CompiledModule, addr: AccountAddress) -> AddressIdentifierIndex {
```
nit: why not spell address and identifier fully?

vgao1996 replied:
Probably just personal preference -- found the names getting a little bit too long

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+/// - A local state store to allow new changes to be stacked on top of the remote state.
+struct SimulationStateView<'a, S> {
+    remote: &'a S,
+    states: Mutex<HashMap<StateKey, Option<StateValue>>>,
```
nit: maybe use `parking_lot::Mutex` to avoid unwraps on poisoning and just release the lock?

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+
+/// Makes `aptos_framework::create_signer::create_signer` a public function so that
+/// it can be used freely in the simulation.
+fn patch_create_signer(state_view: &SimulationStateView<impl StateView>) -> Result<()> {
```
I like this patching and wondering if we can use these mocks beyond this tool to test private functions or hard to test things, like prologue or epilogue in AptosVM. There, we can "inject code in prologue/epilogue" to ensure it gets published and we can control it/assert.

vgao1996 replied:
Like this idea. I've actually been thinking that if we come up with a way to allow these mock versions to be written in Move, they could have a lot more use.

CFG-cased conditional compilation would be sufficient
```
#[cfg(mock)]
fun do_something(..) { .. }

#[cfg(not(mock))]
fun do_something(..) { .. }
```
but it can be a bit wasteful as we have to recompile everything. (May not be a real problem though)

Even better I would love to have a way to write the mock version separately, and only compile that function against the original code.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+        .get_module(&MODULE_ID_CREATE_SIGNER)?
+        .ok_or_else(|| anyhow!("module create_signer does not exist"))?;
+
+    let mut m = CompiledModule::deserialize(&blob)?;
```
I think you should fetch features first to get deserializer config, otherwise deserialization might diverge in behaviour from the actual on-chain

vgao1996 replied:
Good catch.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+        .ok_or_else(|| anyhow!("failed to locate `fun create_signer`"))?;
+    func_def.visibility = Visibility::Public;
+
+    move_bytecode_verifier::verify_module(&m)?;
```
Do we want to run metadata passes (from AptosVM) here as well?

vgao1996 replied:
No, I'm calling the verifier just to some basic sanity check, making sure that the manually-crated bytecode is correct. Will add a more detailed error explanation to make the intent clearer.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+        .get_module(&MODULE_ID_APTOS_GOVERNANCE)?
+        .ok_or_else(|| anyhow!("module create_signer does not exist"))?;
+
+    let mut m = CompiledModule::deserialize(&blob)?;
```
Same here: you should fetch features first and create a deserializer config.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+        &mut m,
+        MODULE_ID_CREATE_SIGNER.address,
+        MODULE_ID_CREATE_SIGNER.name.clone(),
+        Identifier::new("create_signer").unwrap(),
```
nit: we can also have a const for `create_signer` function name? We already do this for module id.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+    let code = func_def.code.as_mut().ok_or_else(|| {
+        anyhow!("`fun resolve_multi_step_proposal` must have a Move-defined body")
+    })?;
+    code.code = vec![
```
Hacky but I am not sure there is another way around :)

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+    let balance = 100 * 1_0000_0000; // 100 APT
+    let account = AccountData::new_from_seed(&mut rng, balance, 0);
+    state_view.apply_write_set(account.to_writeset());
+    // TODO: should update coin info
```
What does this mean? Do you mean the total supply?

vgao1996 replied:
Yes, the total supply.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+    let txn_gas_params = &mut gas_params.vm.txn;
+    // Use the alternative limits for governance proposals
+    // TODO: In the future, consider adding the execution hashes of the scripts to the approval list.
+    txn_gas_params.max_execution_gas = txn_gas_params.max_execution_gas_gov;
```
Doesn't your recent change to automatically bump these numbers to `_gov` if it is a governance proposal work here?

vgao1996 replied:
This is one of the more ugly parts of the current implementation. 

> Doesn't your recent change to automatically bump these numbers to _gov if it is a governance proposal work here?

Yes, but the alt limits will only kick in if the script has its hash added to the list of approved execution hashes, which gets skipped by the mock version of `resolve_multi_step_proposal` as well.

It's possible for us to manually add it in Rust and I tried it, but there are some complexities involved.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,500 @@
+            bail!("failed to execute governance script: {}", script_name)
+        }
+
+        let (write_set, _events) = txn_output.into();
```
Recently we have been discussing with @runtian-zhou that we should always reconfig on governance proposal, especially if it is a framework upgrade (it is not the case now). I expect this simulation tool will be used a lot, should we check that events issue a new epoch event? Asserting `AptosVM::should_restart_execution` sufficies and will be a good place to catch problems.

vgao1996 commented:
Made a major update to the PR
- @georgemitenkov I addressed all your comments and added the `should_restart_execution` check you requested.
- @runtian-zhou I implemented the check you requested, ensuring that the last script cannot have a next execution hash.
- Also fixed two bugs
  - The warm vm cache is now flushed every time we execute a script, so that we always load the latest framework code cc @msmouse 
  - The patching of the framework functions is also done every time we execute a script -- this is needed in case the framework has been overwritten by a previous script.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,596 @@
+            &log_context,
+        );
+        // We require all governance scripts to trigger reconfiguration so check it here.
+        if AptosVM::should_restart_execution(vm_output.events()) {
```
Should it be the negation of this?

vgao1996 replied:
Good catch.

However looks like this would not be an easy fix. With dkg, reconfiguration is started by the script but may not actually happen until the next epoch. 

I guess I'll have to remove this check for now, and we'll need to discuss what the proper solution might be. I'm thinking that maybe `reconfiguration_with_dkg::try_start` should emit its own event indicating this.

georgemitenkov commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,596 @@
+        let script_name = script_path.file_name().unwrap().to_string_lossy();
+        println!("    {}", script_name);
+
+        // Create a new VM to ensure the loader is clean.
```
Have you checked this? Because I think warm vm cache does load PackageMetadata for core packages to see if it has changed... or this is the reason, we patch code and package metadata is the same?

perryjrandall commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,601 @@
+            "--framework-local-dir",
+            &aptos_framework_path().to_string_lossy(),
+            "--skip-fetch-latest-git-deps",
+        ])?;
```
for the ? operator here, I would suggest using Anyhow::Context which can give a little more info on what we were doing / why we failed in case of failure

Currently a lot of our tooling does not use this and since it also does not panic we are left completely without a stack trace although this is pretty straightforward

perryjrandall reviewed (approved):
This is fucking magical, a really important tool for release verification <3

runtian-zhou commented on aptos-move/aptos-release-builder/src/main.rs:
```diff
@@ -134,6 +187,13 @@ async fn main() -> anyhow::Result<()> {
+            network,
+            proposal_dir,
+        } => {
+            simulate_multistep_proposal(network.to_url()?, &proposal_dir).await?;
```
Yea agreed. The CI job will most likely just take a yaml file/

vgao1996 replied:
Okay, I did it in a slightly different way -- you can now add an optional argument `--simulate <network>` to the `generate-proposals` command to do simulation after generation. Does you guys think this would work?

I'm thinking that if I don't do this, then I'll need to add `release_config` as an optional argument to the simulate command. Which one do you think would have the better UX?

runtian-zhou replied:
Optional sounds reasonable to me. You can spawn a temp folder for this.

vgao1996 replied:
Sorry if I didn't make my question clear. Was asking you guys whether you prefer having
```
aptos-release-builder generate-proposals .. --simulate mainnet
```
or
```
aptos-release-builder simulate-multistep-proposal .. --release-config data/release.yaml
```

perryjrandall replied:
I think the --simulate works well is pretty intuitive

runtian-zhou reviewed (commented):
I think the overall logic makes sense to me. The patching logic looked ugly but I don't see a way round for now. We still need a separate PR in the aptos-network to use this command btw.

runtian-zhou commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,601 @@
+        state_view.apply_write_set(write_set);
+    }
+
+    println!("All scripts succeeded!");
```
We need to check there's no pending script hash there. Can you add the check or at least add a todo here? I think it would be important for our release safety.

vgao1996 replied:
I think I've implemented this check, no?

runtian-zhou replied:
Oh nvm. I thought you didn't.

vgao1996 commented on aptos-move/aptos-release-builder/src/simulate.rs:
```diff
@@ -0,0 +1,601 @@
+            })?;
+
+            code.code.clear();
+            if forbid_next_execution_hash {
+                // If it is needed to forbid a next execution hash, inject additional Move
+                // code at the beginning that aborts with a magic number if the vector
+                // representing the hash is not empty.
+                //
+                //     if (!vector::is_empty(&next_execution_hash)) {
+                //         abort MAGIC_FAILED_NEXT_EXECUTION_HASH_CHECK;
+                //     }
+                //
+                // The magic number can later be checked in Rust to determine if such violation
+                // has happened.
+                code.code.extend([
+                    ImmBorrowLoc(2),
+                    VecLen(sig_u8_idx),
+                    LdU64(0),
+                    Eq,
+                    BrTrue(7),
+                    LdU64(MAGIC_FAILED_NEXT_EXECUTION_HASH_CHECK),
+                    Abort,
+                ]);
+            }
```
@runtian-zhou here I check if the next execution hash is empty, and if so, abort with a magic number

runtian-zhou replied:
amazing!

vgao1996 commented:
@perryjrandall I've made some updates to the PR
- Fixed the bug that caused the gas schedule hash to change, by adding script hashes to the approved list properly.
- Added `.context(..)` and `.with_context(..)` in many places to improve error reporting
- Added a helper to modify on-chain config

Additionally I've also created this issue https://github.com/aptos-labs/aptos-core/issues/14044 for tracking follow-up items that I don't plan to address immediately.

perryjrandall commented on aptos-move/aptos-release-builder/src/main.rs:
```diff
@@ -34,6 +73,23 @@ pub enum Commands {
+    /// Simulate a multi-step proposal on the specified network, using its current states.
+    /// The simulation will execute the governance scripts, as if the proposal is already
+    /// approved.
+    SimulateMultiStepProposal {
```
Can we just call this simulate?

We should deprecate the "single step" proposal so everything should eventually just be "multi step" 

And if everything is multi step then nothing is :P

perryjrandall commented on aptos-move/aptos-release-builder/src/main.rs:
```diff
@@ -34,6 +73,23 @@ pub enum Commands {
+    SimulateMultiStepProposal {
+        /// Path to the directory storing the scripts (steps) belonging to the proposal.
+        #[clap(short, long)]
+        proposal_dir: PathBuf,
```
Can we just pass the output dir herre instead of the specific proposal dir?

So right now when u generate you make the output/ folder with many proposal folders potentially

This means you have to call simulate on each one which might be intentional but is a bit annoying you cannot just pass the same high level path and have this figure it out

vgao1996 commented:
Update the PR again.
- @perryjrandall I've renamed the command to "simulate" as you requested. Now it also searches the whole directory recursively, so you can pass in the output dir or any of its sub directories containing the proposals.
- @georgemitenkov  I had to change the way I inject `create_signer` cuz my previous implementation broke module compatibility. This results in an extra flag being passed into the VM, which is a bit ugly, but I guess this is something we can live with and refactor later.

Given that the PR is rather polished right now I'll move forward to landing. If there are additional feature requests I'll address them separately.
